src/main/scala/sigmastate/interpreter/Interpreter.scala

```diff
@@ -313,31 +316,62 @@ trait Interpreter {
     * @return
     */
   def reduceToCrypto(context: CTX, exp: Value[SBoolean.type]): Try[ReductionResult] = Try {
-    require(new Tree(exp).nodes.length < CostTable.MaxExpressions,
-      s"Too long expression, contains ${new Tree(exp).nodes.length} nodes, " +
-        s"allowed maximum is ${CostTable.MaxExpressions}")
-
-    // Substitute Deserialize* nodes with deserialized subtrees
-    // We can estimate cost of the tree evaluation only after this step.
-    val substRule = strategy[Value[_ <: SType]] { case x => substDeserialize(context, x) }
-
-    val substTree = everywherebu(substRule)(exp) match {
-      case Some(v: Value[SBoolean.type]@unchecked) if v.tpe == SBoolean => v
-      case x => throw new Error(s"Context-dependent pre-processing should produce tree of type Boolean but was $x")
-    }
-
-    val cost = substTree.cost(context)
-    if (cost > maxCost) {
-      throw new Error(s"Estimated expression complexity $cost exceeds the limit $maxCost in $substTree")
+    // todo run deserializations first
+    //    // Substitute Deserialize* nodes with deserialized subtrees
+    //    // We can estimate cost of the tree evaluation only after this step.
+    //    val substRule = strategy[Value[_ <: SType]] { case x => substDeserialize(context, x) }
+    //
+    //    val substTree = everywherebu(substRule)(exp) match {
+    //      case Some(v: Value[SBoolean.type]@unchecked) if v.tpe == SBoolean => v
+    //      case x => throw new Error(s"Context-dependent pre-processing should produce tree of type Boolean but was $x")
+    //    }
+
+    // todo get env
+    val env = Map[String, Any]()
+    val IR.Tuple(calcF, costF, sizeF) = doCosting[SType#WrappedType](env, exp)
+    require(IR.verifyCostFunc(costF).isSuccess)
+    require(IR.verifyIsValid(calcF).isSuccess)
+    // check cost
+    val costFun = IR.compile[SInt.type](IR.getDataEnv, costF)
+    val IntConstant(estimatedCost) = costFun(context.toSigmaContext(IR))
+    if (estimatedCost > maxCost) {
+      throw new Error(s"Estimated expression complexity $estimatedCost exceeds the limit $maxCost in $exp")
     }
-
+    // check calc
+    val valueFun = IR.compile[SBoolean.type](IR.getDataEnv, calcF.asRep[IR.Context => SBoolean.WrappedType])
+    //    val res = valueFun(sigmaContext) match {
+    //      case Constant(inVal, _) => inVal
+    //      case v => v
+    //    }
+    val value = valueFun(context.toSigmaContext(IR))
     // After performing deserializations and checking cost of the resulting tree, both the prover
     // and the verifier are evaluating the tree by applying rewriting rules, until no rules trigger during tree
     // traversal.
-    val res = reduceUntilConverged(context, substTree)
-    res -> cost
+    val res = reduceUntilConverged(context, value)
```

Comment:
@greenhat we don't need this anymore, `value` above will already have either boolean or SigmaBoolean (including CAND, COR, etc)